### PR TITLE
core1.0: specify protocol negotiation

### PR DESCRIPTION
**Pull request by majewsky**
We should start using PRs instead of just waiting for the next physical meeting.

### CODE REVIEW

File: spec/core/1.0.md [hunk 1]
```diff
@@ -191,7 +191,7 @@ A message is **invalid** if...
 
 - the size of the bytestring encoding the message exceeds the recipient's maximum message size (see sections 6.1 and 6.2), usually 4096 bytes, or
 
-- the message type name is unknown or its use has not been negotiated with the recipient (see section 4).
+- the message type name is unknown or its use has not been agreed on by the recipient (see section 4).
```
Comment (laerling) [inline spec/core/1.0.md]: From my understanding the server MUST agree to the use of any message type (i. e. the module in which the message type is defined), and the client can not disagree to usage of specific message types whose module already has been agreed on, thus the complete authority over what message types can be used is with the server.
Thus, my proposal:
```
(...) has not been agreed on by the the server (see section 4).
```

File: spec/core/1.0.md [hunk 2]
```diff
@@ -266,6 +266,10 @@ This requirement does not apply when the release is only a draft or pre-release
 
 **Rationale:** This follows the basic notion of [semantic versioning](http://semver.org/spec/v2.0.0.html), albeit massively simplified to suit the usecase of module specifications.
 
+Each module specification whose minor version number is bigger than 0 SHOULD indicate which parts of it have been added or changed compared to previous specifications of that module with the same major version number.
+
+**Rationale:** Implementors should not have to `diff` specification documents manually.
+
 ### 3.2. Compatbility across versions
```
Comment (laerling) [inline spec/core/1.0.md]: Typo: Compat_i_bility

File: spec/core/1.0.md [hunk 3]
```diff
@@ -288,19 +300,106 @@ The name of any unofficial module MUST start with a leading underscore and SHOUL
...
+- Required capabilities: none
+- Number of arguments: one or more
+
+Each argument MUST be a bareword.
+Quoted strings are not allowed in this message.
+
+*Rationale:* The syntax of the `want` message is deliberately minimized because of the compatibility requirements in section 3.2.
+
+The `want` message informs the server of modules and capabilities that the client wants to use.
+Each argument is a module name as accepted by `<module-name>`, or a capability name as accepted by `<message-type>`.
```
Comment (laerling) [inline spec/core/1.0.md]: The use of `<message-type>` allows for the want messages `(want want)` and `(want have)`.
I think it should be defined explicitely if those messages
- are to be answered with `(have want)` and `(have have)` respectively,
- are to be ignored by the server,
- or are invalid and forbidden.

File: spec/core/1.0.md [hunk 4]
```diff
@@ -288,19 +300,106 @@ The name of any unofficial module MUST start with a leading underscore and SHOUL
...
+For each argument that is a capability name, the server MUST already have agreed to using the module that defines this capability, or one of the arguments before this argument MUST be that module's name.
```
Comment (laerling) [inline spec/core/1.0.md]: "or one of the arguments before this argument MUST be that module's name."
I needed to read this sentence a couple of times before understanding this part.
Maybe write "(...)defines this capability. Alternatively one of the arguments that comes before this argument MUST be that module's name."

File: spec/core/1.0.md [hunk 5]
```diff
@@ -288,19 +300,106 @@ The name of any unofficial module MUST start with a leading underscore and SHOUL
...
 
-TODO allow only bareword atoms here; server may not "have" a capability without its module
+- Directionality: server to client
+- Required capabilities: none
+- Number of arguments: zero or more
+
+Each argument MUST be a bareword.
+Quoted strings are not allowed in this message.
+
+*Rationale:* Same as above.
```
Comment (laerling) [inline spec/core/1.0.md]: Reference explicitely, e. g.: "Same as rationale in section 4.1"

File: spec/core/1.0.md [hunk 6]
```diff
@@ -288,19 +300,106 @@ The name of any unofficial module MUST start with a leading underscore and SHOUL
...
+Upon receiving a valid `want` message, the server MUST reply with a `have` message to indicate which modules and capabilities in that `have` message it agrees to.
+If it does not agree to any of the modules and capabilities requested, it must reply with a `have` message with zero arguments.
+
+*Rationale:* A client will typically wait for the `have` reply because the validity of subsequent messages depends on it. A missing `have` message might therefore stall the client forever.
+
+Each argument in the `have` message corresponds to an argument in the `want` message, except that modules and capabilities to which the server does not agree are omitted.
+This particularly means that the argument list of the `have` message MUST preserve the order of arguments in the `want` message.
+
+For `want` arguments that are capability names, the corresponding `have` argument is the same string, the capability name.
+A server SHALL NOT agree to using a capability if it has not previously agreed to using the module defining said capability, unless it agrees to using the module in the same `have` message.
```
Comment (laerling) [inline spec/core/1.0.md]: Proposal:
Explicitely state that the module that is agreed on in the `have` message MUST stand before any capability that it provides and that appears in the same `have` message.

---
Rationale:
This logically follows from the statements:
- At line 324: "(...) or one of the arguments before this argument MUST be that module's name."
- At line 354: "This particularly means that the argument list of the `have` message MUST preserve the order of arguments in the `want` message."

**Review by laerling — COMMENTED** [review]
I agree to using PRs instead of just waiting for the next physical meeting.

File: spec/core/1.0.md (hunk 1 above)
Comment (majewsky) [reply spec/core/1.0.md]: Yes, oversight of mine.

File: spec/core/1.0.md (hunk 3 above)
Comment (majewsky) [reply spec/core/1.0.md]: Will add an exception. `({want,have} {want,have})` is not useful as these messages MUST be implemented by everybody.

File: spec/core/1.0.md (hunk 4 above)
Comment (majewsky) [reply spec/core/1.0.md]: ok

File: spec/core/1.0.md (hunk 5 above)
Comment (majewsky) [reply spec/core/1.0.md]: ok

File: spec/core/1.0.md (hunk 6 above)
Comment (majewsky) [reply spec/core/1.0.md]: I added a clarifying statement, see line 356.

**Comment by majewsky** [comment]
I have incorporated all remarks. If this is good to go for now from your perspective, please approve the PR or write a short notice.

**Review by laerling — APPROVED** [review]

**Pull request merged**